File: Theseus/Plotter/Plotter.cs
using Theseus.Data;

namespace Theseus.Plotter
{
	public static class Plotter
	{
		public static byte[][] Plot(Maze maze)
		{
			byte[][] plot = new byte[maze.Size.X * 2 + 1][];
			for (int x = 0; x < maze.Size.X * 2 + 1; ++x)
			{
				plot[x] = new byte[maze.Size.Y * 2 + 1];
				for (int y = 0; y < maze.Size.Y * 2 + 1; ++y)
				{
					if (y % 2 == 0) // Horizontal wall line
					{
						if (y == maze.Size.Y * 2) // Last horizontal line
						{
							plot[x][y] = 1;
						}
						else if (x % 2 == 0) // Vertical wall line
						{
							// Corner points are always filled
							plot[x][y] = 1;
						}
						else // Vertical room line
						{
							// North-south passages
							Cell cell = maze.Cell((x - 1) / 2, y / 2);
							if (cell.NorthOpen)
								plot[x][y] = 0;
							else
								plot[x][y] = 1;
						}
					}
					else // Horizontal room line
					{
						if (x == maze.Size.X * 2) // Last vertical line
						{
							plot[x][y] = 1;
						}
						else if (x % 2 == 0) // Vertical wall line
						{
							// East-west passages
							Cell cell = maze.Cell(x / 2, (y - 1) / 2);
							if (cell.WestOpen)
								plot[x][y] = 0;
							else
								plot[x][y] = 1;
						}
						else // Vertical room line
						{
							// Rooms
							plot[x][y] = 0;
						}
					}
				}
			}

			int plotX;
			int plotY;

			// Entrance
			plotX = maze.Location(maze.Entrance).X * 2 + 1;
			plotY = maze.Location(maze.Entrance).Y * 2 + 1;

			if (maze.Entrance.North == null)
			{
				plot[plotX][plotY - 1] = 2;
			}
			else if (maze.Entrance.South == null)
			{
				plot[plotX][plotY + 1] = 2;
			}
			else if (maze.Entrance.West == null)
			{
				plot[plotX - 1][plotY] = 2;
			}
			else if (maze.Entrance.East == null)
			{
				plot[plotX + 1][plotY] = 2;
			}

			// Exit
			plotX = maze.Location(maze.Exit).X * 2 + 1;
			plotY = maze.Location(maze.Exit).Y * 2 + 1;

			if (maze.Exit.North == null)
			{
				plot[plotX][plotY - 1] = 3;
			}
			else if (maze.Exit.South == null)
			{
				plot[plotX][plotY + 1] = 3;
			}
			else if (maze.Exit.West == null)
			{
				plot[plotX - 1][plotY] = 3;
			}
			else if (maze.Exit.East == null)
			{
				plot[plotX + 1][plotY] = 3;
			}

			return plot;
		}
	}
}

File: Core/Plotter/Plotter.cpp
#include "Plotter.h"

#include "data/Maze.h"
#include "data/Cell.h"

namespace Theseus { namespace Core {

Plotter* Plotter::plot(Maze* maze)
{
	Plotter* result = new Plotter();

	size_t width = maze->getSizeWidth() * 2 + 1;
	size_t height = maze->getSizeHeight() * 2 + 1;

	char** plot = new char*[height];

	for (size_t y = 0; y < height; ++y)
	{
		plot[y] = new char[width];
		for (size_t x = 0; x < width; ++x)
		{
			plot[y][x] = (char)Void;

			if (y % 2 == 0) // Horizontal wall line
			{
				if (y == maze->getSizeHeight() * 2) // Last horizontal line
				{
					plot[y][x] = Wall;
				}
				else if (x % 2 == 0) // Vertical wall line
				{
					// Corner points are always filled
					plot[y][x] = Wall;
				}
				else // Vertical room line
				{
					// North-south passages
					Cell* cell = maze->getCell((x - 1) / 2, y / 2);
					if (cell->isOpenNorth())
						plot[y][x] = Space;
					else
						plot[y][x] = Wall;
				}
			}
			else // Horizontal room line
			{
				if (x == maze->getSizeWidth() * 2) // Last vertical line
				{
					plot[y][x] = Wall;
				}
				else if (x % 2 == 0) // Vertical wall line
				{
					Cell* cell = maze->getCell(x / 2, (y - 1) / 2);
					if (cell->isOpenWest())
						plot[y][x] = Space;
					else
						plot[y][x] = Wall;
				}
				else // Vertical room line
				{
					plot[y][x] = Space;
				}
			}
		}
	}

	// Entrance and exit
	for (size_t h = 0; h < maze->getSizeHeight(); ++h)
	{
		for (size_t w = 0; w < maze->getSizeWidth(); ++w)
		{
			Cell* cell = maze->getCell(w, h);
			if (cell != maze->getEntrance() && cell != maze->getExit()) continue;

			size_t plotX = w * 2 + 1;
			size_t plotY = h * 2 + 1;

			if (cell->getNorth() == NULL)
				plotY--;
			else if (cell->getWest() == NULL)
				plotX--;
			else if (cell->getSouth() == NULL)
				plotY++;
			else if (cell->getEast() == NULL)
				plotX++;

			if (cell == maze->getEntrance())
				plot[plotY][plotX] = Entrance;

			else if (cell == maze->getExit())
				plot[plotY][plotX] = Exit;
		}
	}

	result->_width = width;
	result->_height = height;
	result->_plot = plot;
	return result;
}

Plotter::~Plotter()
{
	for (size_t y = 0; y < _height; ++y)
	{
		delete[] _plot[y];
	}

	delete[] _plot;
}

size_t Plotter::getSizeWidth() const
{
	return _width;
}

size_t Plotter::getSizeHeight() const
{
	return _height;
}

char Plotter::getPixel(size_t x, size_t y) const
{
	return _plot[y][x];
}

}}